backend/libs/artemislib/artemislib/logging.py

```diff
@@ -1,29 +1,182 @@
+
+        log_record = self._prune_null_fields(log_record)
+        return json.dumps(log_record)
+
+    def _prune_null_fields(self, records: Dict[str, Any]) -> Dict[str, Any]:
+        """
+        Remove any key with a Null value
+        """
+        return {k: v for k, v in records.items() if v is not None}
 
-LEVEL_MAP = {
-    "CRITICAL": logging.CRITICAL,
-    "ERROR": logging.ERROR,
-    "WARNING": logging.WARNING,
-    "INFO": logging.INFO,
-    "DEBUG": logging.DEBUG,
-    "NOTSET": logging.NOTSET,
-}
+
+class LogRecordFactory:
+    """
+    A custom Factory class to be used when instantiating a LogRecord.
+
+    LogRecord instances are created every time something is logged. They
+    contain all the information for the event being logged. This custom
+    factory, adds the ability to add and remove fields from a LogRecord
+    """
+
+    def __init__(self):
+        self.extra_fields = {}
+
+    def __call__(self, *args, **kwargs) -> LogRecord:
+        record = LogRecord(*args, **kwargs)
+        for key, value in self.extra_fields.items():
+            setattr(record, key, value)
+        return record
+
+    def add_fields(self, **kwargs):
+        self.extra_fields.update(kwargs)
+
+    def remove_fields(self, *args):
+        for key in args:
+            self.extra_fields.pop(key, None)
+
+    def reset_fields(self):
+        self.extra_fields = {}
+
+    def get_current_fields(self) -> Dict[str, Any]:
+        return self.extra_fields
 
 
 class Logger:
-    def __new__(cls, name: str, level=LOG_LEVEL):
-        log = logging.getLogger(name.strip())
+    """
+    Factory Class for generating new Loggers
+
+    Provides options to update log fields that that should persist across all loggers
+    """
+
+    def __new__(cls, name: str, level=LOG_LEVEL, stream=sys.stdout):
+        log = getLogger(name.strip())
         if not log.hasHandlers():
-            console = logging.StreamHandler(sys.stderr)
-            formatter = logging.Formatter(
-                fmt="%(asctime)s %(levelname)-8s [%(name)s] %(message)s", datefmt="[%Y-%m-%dT%H:%M:%S%z]"
-            )
-            console.setFormatter(formatter)
-            log.addHandler(console)
+            json_formatter = JSONFormatter()
+            stdout_handler = StreamHandler(stream)
+            stdout_handler.setFormatter(json_formatter)
+            log.addHandler(stdout_handler)
+
         log.setLevel(LEVEL_MAP.get(level, DEFAULT_LOG_LEVEL))
         return log
+
+    @classmethod
+    def add_fields(cls, **kwargs):
+        record_factory = cls._get_record_factory()
+        record_factory.add_fields(**kwargs)
+
+    @classmethod
+    def remove_fields(cls, *args):
+        record_factory = cls._get_record_factory()
+        record_factory.remove_fields(*args)
+
+    @classmethod
+    def reset_fields(cls):
+        record_factory = cls._get_record_factory()
+        record_factory.reset_fields()
```

Comment:
We should implement a context manager for adding fields and making sure those fields are removed/restored at the end of the block.  The Python logging cookbook has an example where they implement a context manager to adjust logging levels within a block; we can adapt that for saving/restoring fields: https://docs.python.org/3/howto/logging-cookbook.html#using-a-context-manager-for-selective-logging

Needing to make sure calls to `add_fields()` and `remove_fields()` are matched up tends to be brittle and difficult to review.  Something like this would be much easier:

```python
from artemislib.logging import LogFields

for plugin in self.action_details.plugins:
  with LogFields(logger, plugin=plugin):
    # The "plugin" field automatically gets restored at the end of the block
    logger.info( ... )
```

That said, for now, let's not let perfect be the enemy of good and save that for a later enhancement.

Reply:
Created a task to implement this